File: ros_ws/esmini-bridge-autoware/src/ego_init.cpp
#include "ego_init.hpp"

EgoInitNode::EgoInitNode(const rclcpp::NodeOptions &options)
    : Node("EgoInitNode", options) {
    RCLCPP_INFO(this->get_logger(), "EgoInitNode Constructed");
    this->pub_initialpose_ =
        this->create_publisher<PoseWithCovarianceStamped>("/initialpose", 10);

    // sleep 1
    std::this_thread::sleep_for(std::chrono::seconds(1));
    this->run();
}

void EgoInitNode::run() {
    auto msg = PoseWithCovarianceStamped();
    msg.header.stamp = get_clock()->now();
    msg.header.frame_id = "map";
    // msg.header.stamp = this->now();
    msg.pose.pose.position.x = 51.0;
    msg.pose.pose.position.y = -1;
    msg.pose.pose.position.z = 0;
    msg.pose.pose.orientation.x = 0;
    msg.pose.pose.orientation.y = 0;
    msg.pose.pose.orientation.z = 0;
    msg.pose.pose.orientation.w = 1;
    this->pub_initialpose_->publish(msg);
    RCLCPP_INFO(this->get_logger(), "EgoInitNode Published");
}

#include <rclcpp_components/register_node_macro.hpp>
RCLCPP_COMPONENTS_REGISTER_NODE(EgoInitNode)

File: ros_ws/src/esmini-bridge-autoware/src/Autoware/EgoInit.py
import time
import rclpy
from rclpy.node import Node
from tf_transformations import quaternion_from_euler

from geometry_msgs.msg import PoseWithCovarianceStamped, PoseStamped

from autoware_adapi_v1_msgs.srv import ChangeOperationMode


class EgoInit(Node):
    def __init__(self):
        super().__init__("egoinit")
        self.logger = self.get_logger()
        self._init_publisher = self.create_publisher(
            PoseWithCovarianceStamped, "/initialpose", 10
        )
        self._goal_publisher = self.create_publisher(
            PoseStamped, "/planning/mission_planning/goal", 10
        )
        self._switch_mode_client = self.create_client(
            ChangeOperationMode,
            "/api/operation_mode/change_to_autonomous",
        )

    def publish_initial_pose(self, x, y, h):
        self.logger.info("heading: " + str(h))
        msg = PoseWithCovarianceStamped()
        msg.header.stamp = self.get_clock().now().to_msg()
        msg.header.frame_id = "map"
        msg.pose.pose.position.x = x
        msg.pose.pose.position.y = y
        quat = quaternion_from_euler(0, 0, h)
        msg.pose.pose.orientation.x = quat[0]
        msg.pose.pose.orientation.y = quat[1]
        msg.pose.pose.orientation.z = quat[2]
        msg.pose.pose.orientation.w = quat[3]
        self._init_publisher.publish(msg)

    def publish_goal_pose(self):
        msg = PoseStamped()
        msg.header.stamp = self.get_clock().now().to_msg()
        msg.header.frame_id = "map"
        msg.pose.position.x = 267.456665
        msg.pose.position.y = -92.682770
        msg.pose.orientation.z = -0.712069
        msg.pose.orientation.w = 0.702110
        self.logger.info(f"Publish goal {msg}")
        self._goal_publisher.publish(msg)

    def switch_auto_mode(self):
        self.logger.info(f"Wait Autoware a second to plan a route")
        time.sleep(1)
        done = False
        while not done:
            while not self._switch_mode_client.wait_for_service(timeout_sec=1.0):
                self.get_logger().info("service not available, waiting again...")
            req = ChangeOperationMode.Request()
            future = self._switch_mode_client.call_async(req)
            rclpy.spin_until_future_complete(self, future)
            self.logger.debug(
                f"Calling {self._switch_mode_client.srv_name} got result {future.result()}"
            )
            if future.result().status.success:
                self.logger.info("Switch mode success")
                done = True
            else:
                self.logger.info(
                    f"Switch mode failed with msg: {future.result().status.message}, retrying..."
                )
                time.sleep(1)
